runtime/security.go

```diff
@@ -166,172 +103,199 @@ func (p *securityEngine) resolveMetricsViewSecurity(instanceID, environment stri
 	}
 
-	resolved := &ResolvedMetricsViewSecurity{Access: true}
 	templateData := rillv1.TemplateData{
 		Environment: environment,
 		User:        attributes,
 		Self:        rillv1.TemplateResource{Meta: mv.Meta},
 	}
 
-	for _, policy := range policies {
-		if policy == nil {
+	// Apply rules
+	var access *bool
+	var rowFilter string
+	var queryFilter *runtimev1.Expression
+	var fieldAccess map[string]bool
+	for _, rule := range rules {
+		if rule == nil {
 			continue
 		}
 
-		if policy.Access != "" {
-			accessExpr, err := rillv1.ResolveTemplate(policy.Access, templateData)
-			if err != nil {
-				return nil, err
+		switch r := rule.Rule.(type) {
+		case *runtimev1.SecurityRule_Access:
+			// Determine if the rule should be applied
+			apply := true
+			if r.Access.Condition != "" {
+				expr, err := rillv1.ResolveTemplate(r.Access.Condition, templateData)
+				if err != nil {
+					return nil, err
+				}
+				res, err := rillv1.EvaluateBoolExpression(expr)
+				if err != nil {
+					return nil, err
+				}
+				apply = res
 			}
-			access, err := rillv1.EvaluateBoolExpression(accessExpr)
-			if err != nil {
-				return nil, err
+			if !apply {
+				continue
 			}
 
-			resolved.Access = resolved.Access && access
-		} else {
-			resolved.Access = false
-		}
-
-		if policy.RowFilter != "" {
-			filter, err := rillv1.ResolveTemplate(policy.RowFilter, templateData)
-			if err != nil {
-				return nil, err
-			}
-
-			if resolved.RowFilter == "" {
-				resolved.RowFilter = filter
+			// Explicit denies take precedence
+			if access == nil {
+				access = &r.Access.Allow
 			} else {
-				resolved.RowFilter = fmt.Sprintf("(%s) AND (%s)", resolved.RowFilter, filter)
+				tmp := *access && r.Access.Allow
+				access = &tmp
 			}
-		}
-
-		if policy.QueryFilter != nil {
-			if resolved.QueryFilter == nil {
-				resolved.QueryFilter = policy.QueryFilter
-			} else {
-				resolved.QueryFilter = expressionpb.And([]*runtimev1.Expression{resolved.QueryFilter, policy.QueryFilter})
+		case *runtimev1.SecurityRule_FieldAccess:
+			// As soon as we see a field access rule, we set fieldAccess, which entails an implicit deny for all fields not mentioned
+			if fieldAccess == nil {
+				fieldAccess = make(map[string]bool)
 			}
-		}
 
-		var include, exclude []string
-		var excludeAll bool
-		seen := map[string]bool{}
-
-		for _, inc := range policy.Include {
-			cond, err := rillv1.ResolveTemplate(inc.Condition, templateData)
-			if err != nil {
-				return nil, err
+			// Determine if the rule should be applied
+			apply := true
+			if r.FieldAccess.Condition != "" {
+				expr, err := rillv1.ResolveTemplate(r.FieldAccess.Condition, templateData)
+				if err != nil {
+					return nil, err
+				}
+				res, err := rillv1.EvaluateBoolExpression(expr)
+				if err != nil {
+					return nil, err
+				}
+				apply = res
 			}
-			incCond, err := rillv1.EvaluateBoolExpression(cond)
-			if err != nil {
-				return nil, err
+			if !apply {
```

Comment:
if condition is `false` or empty then fieldAccess will be empty and all fields will be allowed, should we explicitly ask what action they want here allow or deny ? also I think this is not backwards compatible behaviour. May be we enforce specifying an `action` in the YAML then we use the `allow` field here, I hope I am not mixing things up.

Reply:
Actually it is the reverse, as soon as a field access rule is seen, even if it is false, all fields are implicitly denied until explicitly allowed. The only case for implicit allow is if there are no field access rules at all (i.e. no include/exclude properties). I believe this should be backwards compatible?

See this code: https://github.com/rilldata/rill/blob/36573348772a602dabf3ef5b0101323ce4a5393c/runtime/security.go#L148-L153

Reply:
ok got it, makes sense.